src/peft/tuners/lora/tp_layer.py

```diff
@@ -0,0 +1,145 @@
+from typing import Any
+
+import torch
+import torch.nn as nn
+import torch.nn.init as init
+
+from .layer import LoraLayer
+
+
+class LoraParallelLinear(nn.Module, LoraLayer):
+    """
+    When the target layer parallel_linear is RowParallelLinear, in order to keep the input and output shapes
+    consistent, we need to split the lora matrix A into rows, and the lora_B at this time should be a complete linear
+    layer; In the same way, when the target layer is ColumnParallelLinear, we perform column segmentation on lora_B,
+    while lora_A is still a complete linear layer.
+    """
+
+    def __init__(
+        self,
+        base_layer,
+        adapter_name: str,
+        backend,
+        r: int = 0,
+        lora_alpha: int = 1,
+        lora_dropout: float = 0.0,
+        fan_in_fan_out: bool = False,
+        init_lora_weights: bool = True,
+        **kwargs,
+    ):
+        super().__init__()
+        LoraLayer.__init__(self, base_layer=base_layer)
+
+        self.backend = backend
+        self.is_paralle_a = isinstance(base_layer, backend.RowParallelLinear)
+        self.fan_in_fan_out = fan_in_fan_out
+        self._active_adapter = adapter_name
+
+        megatron_config = kwargs["megatron_config"]
+        parallel_linear_kwargs = {"megatron_config": megatron_config}
+        if hasattr(megatron_config, "init_method"):
+            parallel_linear_kwargs["init_method"] = megatron_config.init_method
+        if isinstance(base_layer, self.backend.RowParallelLinear):
+            parallel_linear_kwargs["input_is_parallel"] = base_layer.input_is_parallel
+        else:
+            parallel_linear_kwargs["gather_output"] = base_layer.gather_output
```

Comment:
For these 3 arguments and for `megatron_config`, is there really a point in putting them inside of `parallel_linear_kwargs`? They are unpacked anyway when passed to `self.update_layer`, so it just seems like an unnecessary indirection. I think it would be easier to just pass all of them explicitly and not have any `**kwargs`.

Reply:
> For these 3 arguments and for `megatron_config`, is there really a point in putting them inside of `parallel_linear_kwargs`? They are unpacked anyway when passed to `self.update_layer`, so it just seems like an unnecessary indirection. I think it would be easier to just pass all of them explicitly and not have any `**kwargs`.

Yes, it is indeed clearer now.